src/firecracker/src/main.rs

```diff
@@ -35,51 +39,58 @@ const DEFAULT_INSTANCE_ID: &str = "anonymous-instance";
 const FIRECRACKER_VERSION: &str = env!("FIRECRACKER_VERSION");
 const MMDS_CONTENT_ARG: &str = "metadata";
 
-#[cfg(target_arch = "aarch64")]
-/// Enable SSBD mitigation through `prctl`.
-pub fn enable_ssbd_mitigation() {
-    // Parameters for `prctl`
-    // TODO: generate bindings for these from the kernel sources.
-    // https://elixir.bootlin.com/linux/v4.17/source/include/uapi/linux/prctl.h#L212
-    const PR_SET_SPECULATION_CTRL: i32 = 53;
-    const PR_SPEC_STORE_BYPASS: u64 = 0;
-    const PR_SPEC_FORCE_DISABLE: u64 = 1u64 << 3;
+#[derive(Debug, thiserror::Error)]
+enum MainError {
+    #[error("Failed to register signal handlers: {0}")]
+    RegisterSignalHandlers(#[source] utils::errno::Error),
+    #[error("Arguments parsing error: {0} \n\nFor more information try --help.")]
+    ParseArguments(#[from] utils::arg_parser::Error),
+    #[error("When printing Snapshot Data format: {0}")]
+    PrintSnapshotDataFormat(#[from] SnapshotVersionError),
+    #[error("Invalid value for logger level: {0}.Possible values: [Error, Warning, Info, Debug]")]
+    InvalidLogLevel(LoggerConfigError),
+    #[error("Could not initialize logger: {0}")]
+    LoggerInitialization(LoggerConfigError),
+    #[error("Could not initialize metrics: {0:?}")]
+    MetricsInitialization(MetricsConfigError),
+    #[error("Seccomp error: {0}")]
+    SeccompFilter(FilterError),
+    #[error("RunWithApiError error: {0}")]
+    RunWithApi(ApiServerError),
+    #[error("RunWithoutApiError error: {0}")]
+    RunWithoutApiError(RunWithoutApiError),
+}
 
-    // SAFETY: Parameters are valid since they are copied verbatim
-    // from the kernel's UAPI.
-    // PR_SET_SPECULATION_CTRL only uses those 2 parameters, so it's ok
-    // to leave the latter 2 as zero.
-    let ret = unsafe {
-        libc::prctl(
-            PR_SET_SPECULATION_CTRL,
-            PR_SPEC_STORE_BYPASS,
-            PR_SPEC_FORCE_DISABLE,
-            0,
-            0,
-        )
-    };
+impl From<MainError> for ExitCode {
+    fn from(value: MainError) -> Self {
+        let exit_code = match value {
+            MainError::ParseArguments(_) => FcExitCode::ArgParsing,
+            MainError::InvalidLogLevel(_) => FcExitCode::BadConfiguration,
+            MainError::RunWithApi(ApiServerError::MicroVMStoppedWithoutError(code)) => code,
+            MainError::RunWithApi(ApiServerError::MicroVMStoppedWithError(code)) => code,
+            _ => FcExitCode::GenericError,
+        };
 
-    if ret < 0 {
-        let last_error = std::io::Error::last_os_error().raw_os_error().unwrap();
-        error!(
-            "Could not enable SSBD mitigation through prctl, error {}",
-            last_error
-        );
-        if last_error == libc::EINVAL {
-            error!("The host does not support SSBD mitigation through prctl.");
-        }
+        ExitCode::from(exit_code as u8)
     }
 }
 
-fn main_exitable() -> FcExitCode {
+fn main() -> ExitCode {
```

Comment:
Can we change this to returns a `Result<(), MainError>` like the other binaries?

Reply:
[`std::process::Termination`](https://doc.rust-lang.org/std/process/trait.Termination.html) determines the exit code, so it can return error codes other than 1 if implemented for an error.

Reply:
But what if we did not get an error? If we have `fn main() -> Error` then we can not exit normally without `Error` having some `success` case. But this does not make sense for me.

Reply:
And `-> Result<(), Error>` will return 1 in case of any error. https://doc.rust-lang.org/src/std/process.rs.html#2272

Reply:
It could be implemented like `struct MainReturn(Result<(),MainError>);`

```rust
impl Termination for MainReturn {
    fn report(self) -> ExitCode {
        match self.0 {
            Ok(()) => ExitCode::SUCCESS,
            Err(err) => err.report()
        }
    }
}
```

Reply:
Well.. now you just moved code from main into separate struct, but in nutshell it is same as what I have right now.

Reply:
Yeah, its practically the same.